Source/bmalloc/libpas/src/libpas/pas_report_crash.h

```diff
@@ -0,0 +1,59 @@
+/*
+ * Copyright (c) 2023 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef PAS_REPORT_CRASH_H
+#define PAS_REPORT_CRASH_H
+
+#include "pas_utils.h"
+
+#include "pas_ptr_hash_set.h"
+#include "pas_ptr_hash_map.h"
+#include "pas_root.h"
+
+#if PAS_OS(DARWIN)
+#include <mach/mach_types.h>
+#include <mach/vm_types.h>
+#endif
+
+PAS_BEGIN_EXTERN_C;
+
+#if PAS_OS(DARWIN)
+// Read memory from crashed process.
+typedef void *(*crash_reporter_memory_reader_t)(task_t task, vm_address_t address, size_t size);
+
+// Report sent back to the ReportCrash process.
+typedef struct {
+    const char *error_type;
+    const char *confidence;
+    vm_address_t fault_address;
+    size_t allocation_size;
+} pas_report_crash_pgm_report;
+
+kern_return_t pas_report_crash_extract_pgm_failure(vm_address_t fault_address, mach_vm_address_t pas_dead_root, task_t task, pas_report_crash_pgm_report *report, crash_reporter_memory_reader_t crm_reader);
```

Comment:
I think we need to mark this function as exported (via `BEXPORT` macro) if we want the symbol to be available in `JavaScriptCore.framework` (and not stripped as dead code).

Reply:
I am not sure we want to start taking bmalloc code and requiring for libpas at the moment.

Elliott and I were discussing adding code to JavaScriptCore that would call into libpas.

We hide libpas symbols now if you try to go through linking JSC.

Reply:
Right – as of c885def, JavaScriptCore hides all symbols from libpas (making libpas an effective implementation detail). This was done to better align the symbols exported from JavaScriptCore with the declarations visible in JavaScriptCore/WTF/bmalloc headers. (And _that_ work is in support of InstallAPI, see https://bugs.webkit.org/show_bug.cgi?id=254784.)

Reply:
I suggested that Brandon add SPI to JavaScriptCore that proxies through to this pas call, since that respects the "libpas is implementation-only" principle we are now enforcing via the linker.

Reply:
> I suggested that Brandon add SPI to JavaScriptCore that proxies through to this pas call,

This could also be implemented in bmalloc, since it's _only_ the libpas symbols that are hidden. (bmalloc already has clients outside of JavaScriptCore, such as WebCore and WebKit.)